### PR TITLE
Fix duplicate sequence index for kraken trades

### PR DESCRIPTION
**Pull request by yabirgb**
Fixes duplicate sequence index + upgrades db + fix logic in kraken module

### CODE REVIEW

File: rotkehlchen/db/upgrades/v31_v32.py [hunk 1]
```diff
@@ -86,6 +86,23 @@ def _add_new_tables(cursor: 'Cursor') -> None:
...
+    being used. This function makes the state of the database consistent with the upgraded
+    logic by:
+    - Removing extra row additions
+    - Updating the sequence index to avoid duplicates
+    """
+    cursor.execute("""
+    DELETE FROM history_events where location="B" AND asset="KFEE" AND
+     type="trade" AND subtype=NULL;
+    """)
+    cursor.execute("""
```
[the remark this answers is not in the document]
Comment (LefterisJP) [reply rotkehlchen/db/upgrades/v31_v32.py:100]: > because the check can't be done in sqlite as far as I know 

It can. How do you think I found it out in the first place? I posted it yesterday in discord:

```
SELECT e.event_identifier, e.sequence_index from history_events e JOIN (SELECT event_identifier, sequence_index, COUNT(*) as cnt FROM history_events GROUP BY event_identifier, sequence_index) other ON e.event_identifier = other.event_identifier and e.sequence_index=other.sequence_index WHERE other.cnt > 1;
```

The main question would be what to do if say we (1) run your upgrade change and (2) run this check and the check still gives a result.

If it still gives a result we would have to decide what to do with it. At DB upgrade time we probably would not be able to know. Only "solution" would be an ugly one and it would mean to just assign a random sequence index to any such entries.

Perhaps I am overthinking it. Not sure. I just don't want to have people stuck at failed upgrades. Keep in mind we still don't have the user editing tool for the DB.

[the remark this answers is not in the document]
Comment (yabirgb) [reply rotkehlchen/db/upgrades/v31_v32.py:100]: I was thinking in handling the update and everything from sqlite but you are right, probably can be done. Yes, having people stuck with upgrades would be the real issue

File: rotkehlchen/db/upgrades/v31_v32.py [hunk 2]
```diff
@@ -86,6 +87,52 @@ def _add_new_tables(cursor: 'Cursor') -> None:
...
+    - Removing extra row additions
+    - Updating the sequence index to avoid duplicates
+    - Make sure that no other event has duplicated sequence indexes
+    """
+    cursor.execute("""
+    DELETE FROM history_events where location="B" AND asset="KFEE" AND
+     type="trade" AND subtype=NULL;
+    """)
+    cursor.execute("""
+    UPDATE history_events SET sequence_index = sequence_index + 1 WHERE asset="KFEE"
```
Comment (LefterisJP) [inline rotkehlchen/db/upgrades/v31_v32.py:103]: What if this creates a conflict? Why is this line here?

File: rotkehlchen/db/upgrades/v31_v32.py [hunk 3]
```diff
@@ -86,6 +87,52 @@ def _add_new_tables(cursor: 'Cursor') -> None:
...
+
+    cursor.execute("""
+    SELECT e.event_identifier, e.sequence_index, e.identifier from history_events e JOIN (SELECT event_identifier,
+    sequence_index, COUNT(*) as cnt FROM history_events GROUP BY event_identifier, sequence_index)
+    other ON e.event_identifier = other.event_identifier and e.sequence_index=other.sequence_index
+    WHERE other.cnt > 1;
+    """)  # noqa: E501
+
+    update_sentences = []
+    used_indexes: Dict[str, Set[int]] = defaultdict(set)
```
Comment (LefterisJP) [inline rotkehlchen/db/upgrades/v31_v32.py:115]: names are very bad here and confuse the reader.

(can also use indexes -- but I would prefer if you used indices as plural of index since this is what we use in the codebase)
```suggestion
    eventid_to_indices: Dict[str, Set[int]] = defaultdict(set)
```

File: rotkehlchen/db/upgrades/v31_v32.py [hunk 4]
```diff
@@ -86,6 +87,52 @@ def _add_new_tables(cursor: 'Cursor') -> None:
...
+    for event_identifier, sequence_index, identifier in cursor:
+        last_index = used_indexes.get(event_identifier)
+        if last_index is None:
```
Comment (LefterisJP) [inline rotkehlchen/db/upgrades/v31_v32.py:118]: This confused me so much! The name should be plural. It's a set not an int you are getting here.

```suggestion
        last_indices = used_indexes.get(event_identifier)
        if last_indices is None:
```

File: rotkehlchen/db/upgrades/v31_v32.py [hunk 5]
```diff
@@ -86,6 +87,52 @@ def _add_new_tables(cursor: 'Cursor') -> None:
...
+    """)
+
+    cursor.execute("""
+    SELECT e.event_identifier, e.sequence_index, e.identifier from history_events e JOIN (SELECT event_identifier,
+    sequence_index, COUNT(*) as cnt FROM history_events GROUP BY event_identifier, sequence_index)
+    other ON e.event_identifier = other.event_identifier and e.sequence_index=other.sequence_index
+    WHERE other.cnt > 1;
+    """)  # noqa: E501
+
+    update_sentences = []
```
Comment (LefterisJP) [inline rotkehlchen/db/upgrades/v31_v32.py:114]: It's a list of tuples, not sentences :smile: 

```suggestion
    update_tuples = []
```

File: rotkehlchen/db/upgrades/v31_v32.py [hunk 6]
```diff
@@ -86,6 +87,52 @@ def _add_new_tables(cursor: 'Cursor') -> None:
...
+    DELETE FROM history_events where location="B" AND asset="KFEE" AND
+     type="trade" AND subtype=NULL;
+    """)
+    cursor.execute("""
+    UPDATE history_events SET sequence_index = sequence_index + 1 WHERE asset="KFEE"
+     AND location="B" AND type="trade" AND subtype="fee"
+    """)
+
+    cursor.execute("""
+    SELECT e.event_identifier, e.sequence_index, e.identifier from history_events e JOIN (SELECT event_identifier,
```
Comment (LefterisJP) [inline rotkehlchen/db/upgrades/v31_v32.py:108]: Have you tested that if there is multiple duplicates such as:

```
eventid1, 1
eventid1, 1
eventid2, 3
eventid2, 3
```
they will all be returned?

File: rotkehlchen/exchanges/kraken.py [hunk 7]
```diff
@@ -212,25 +213,27 @@ def history_event_from_kraken(
...
-                asset=asset,
-                balance=Balance(
-                    amount=raw_amount,
-                    usd_value=ZERO,
-                ),
-                notes=notes,
-                event_type=event_type,
-                event_subtype=event_subtype,
-            ))
+            # Make sure to not generate a event for KFEES that is not of type FEE
```
[the remark this answers is not in the document]
Comment (LefterisJP) [reply rotkehlchen/exchanges/kraken.py:216]: ping

File: rotkehlchen/tests/db/test_db_upgrades.py [hunk 8]
```diff
@@ -2427,12 +2427,12 @@ def test_upgrade_db_31_to_32(user_data_dir):  # pylint: disable=unused-argument
...
     result = cursor.execute('SELECT rowid from history_events')
     old_ids = {row[0] for row in result}
-    assert len(old_ids) == 3
+    assert len(old_ids) == 13
     cursor.execute(
         'SELECT subtype from history_events',
     )
     subtypes = [row[0] for row in cursor]
-    assert set(subtypes) == {'staking deposit asset', 'staking receive asset', None}
+    assert set(subtypes) == {'staking deposit asset', 'staking receive asset', None, 'fee'}
```
Comment (LefterisJP) [inline rotkehlchen/tests/db/test_db_upgrades.py:2435]: I would like to also see a test case here for what I mentioned for the DB upgrade and the duplicates existing in the DB before the upgrade and being "fixed" after.

**Review by LefterisJP — COMMENTED** [review]
Left some late night comments

File: rotkehlchen/db/upgrades/v31_v32.py (hunk 6 above)
Comment (yabirgb) [reply rotkehlchen/db/upgrades/v31_v32.py:108]: yes I have tested

```
# The trade
"KRAKEN-REMOTE-ID3",0
"KRAKEN-REMOTE-ID3",1
"KRAKEN-REMOTE-ID3",3
"KRAKEN-REMOTE-ID3",2
"KRAKEN-REMOTE-ID3",3

# Other event
"KRAKEN-REMOTE-ID4",0
"KRAKEN-REMOTE-ID4",1
"KRAKEN-REMOTE-ID4",1
"KRAKEN-REMOTE-ID4",2
"KRAKEN-REMOTE-ID4",2
```

File: rotkehlchen/db/upgrades/v31_v32.py (hunk 2 above)
Comment (yabirgb) [reply rotkehlchen/db/upgrades/v31_v32.py:103]: I just removed it. Wanted to make sure that those are removed but indeed the other logic is enough

File: rotkehlchen/tests/db/test_db_upgrades.py [hunk 9]
```diff
@@ -2469,7 +2469,21 @@ def test_upgrade_db_31_to_32(user_data_dir):  # pylint: disable=unused-argument
         ('1111112', 1595640208, 'external', 'ETH', 'USD', 'buy', '1.5541', '22.1', '3.4', None, None, None),  # noqa: E501
         ('1111113', 1595640208, 'external', 'ETH', 'USD', 'buy', '1.5541', '22.1', None, 'USD', None, None),  # noqa: E501
     ]
-
+    # Check that there are invalid pairs of (event_identifier, sequence_index)
+    base_entries_query = 'SELECT * from history_events WHERE event_identifier="KRAKEN-REMOTE-ID3"'
+    result = cursor.execute(base_entries_query).fetchall()
+    assert len(result) == 5
+    assert len([row[2] for row in result]) == 5
```
Comment (LefterisJP) [inline rotkehlchen/tests/db/test_db_upgrades.py:2476]: Lol man wtf. I had to look at this twice to understand how you use it to check the invalid indices. I thought it was copy paste row with different assertion at first. Probably the worst way to do it performance wise but since it's just a test I am gonna overlook as it's short :)

File: rotkehlchen/tests/db/test_db_upgrades.py [hunk 10]
```diff
@@ -2522,6 +2536,19 @@ def test_upgrade_db_31_to_32(user_data_dir):  # pylint: disable=unused-argument
         ('1111113', 1595640208, 'external', 'ETH', 'USD', 'buy', '1.5541', '22.1', None, None, None, None),  # noqa: E501
     ]
 
+    # Check that sequence indeces are unique for the same event identifier
```
Comment (LefterisJP) [inline rotkehlchen/tests/db/test_db_upgrades.py:2539]:
```suggestion
    # Check that sequence indices are unique for the same event identifier
```

**Review by LefterisJP — APPROVED** [review]
Only some nitpicks so will merge.

Failing tests pass locally and are flaky.

Need to re-check e2e at some point since not sure if they need more love after Isaac's PR.
